src/backend/cpu/mean.cpp

```diff
@@ -0,0 +1,183 @@
+
+template<typename Ti, typename Tw, typename To>
+Array<To> mean(const Array<Ti>& in, const int dim)
+{
+    in.eval();
+
+    dim4 odims = in.dims();
+    odims[dim] = 1;
+    Array<To> out = createEmptyArray<To>(odims);
+    static const mean_dim_func<Ti, Tw, To> mean_funcs[] = { kernel::mean_dim<Ti, Tw, To, 1>()
+                                                           , kernel::mean_dim<Ti, Tw, To, 2>()
+                                                           , kernel::mean_dim<Ti, Tw, To, 3>()
+                                                           , kernel::mean_dim<Ti, Tw, To, 4>()};
+
+    getQueue().enqueue(mean_funcs[in.ndims() - 1], out, 0, in, 0, dim);
+    return out;
+}
+
+template<typename T, typename Tw>
+using mean_weighted_dim_func = std::function<void(Array<T>, const dim_t,
+                                            const Array<T>, const dim_t, const Array<Tw>, const dim_t, const int)>;
+
+template<typename T, typename Tw>
+Array<T> mean(const Array<T>& in, const Array<Tw>& wt, const int dim)
+{
+    in.eval();
+    wt.eval();
+
+    dim4 odims = in.dims();
+    odims[dim] = 1;
+    Array<T> out = createEmptyArray<T>(odims);
+    static const mean_weighted_dim_func<T, Tw> mean_funcs[] = { kernel::mean_weighted_dim<T, Tw, 1>()
+                                                           , kernel::mean_weighted_dim<T, Tw, 2>()
+                                                           , kernel::mean_weighted_dim<T, Tw, 3>()
+                                                           , kernel::mean_weighted_dim<T, Tw, 4>()};
+
+    getQueue().enqueue(mean_funcs[in.ndims() - 1], out, 0, in, 0, wt, 0, dim);
+    return out;
+}
+
+template<typename T, typename Tw>
+T mean(const Array<T>& in, const Array<Tw>& wt)
+{
+    in.eval();
+    wt.eval();
+    getQueue().sync();
+
+    af::dim4 dims = in.dims();
+    af::dim4 strides = in.strides();
+    const T  *inPtr = in.get();
+    const Tw *wtPtr = wt.get();
+
+    kernel::MeanOp<T, Tw> Op(inPtr[0], wtPtr[0]);
+
+    for(dim_t l = 0; l < dims[3]; l++) {
+        dim_t off3 = l * strides[3];
+
+        for(dim_t k = 0; k < dims[2]; k++) {
+            dim_t off2 = k * strides[2];
+
+            for(dim_t j = 0; j < dims[1]; j++) {
+                dim_t off1 = j * strides[1];
+
+                for(dim_t i = 0; i < dims[0]; i++) {
+                    dim_t idx = i + off1 + off2 + off3;
+                    Op(inPtr[idx], wtPtr[idx]);
+                }
+            }
+        }
+    }
+
+    return Op.runningMean;
+}
+
+template<typename Ti, typename Tw, typename To>
+To mean(const Array<Ti>& in)
+{
+    in.eval();
+    getQueue().sync();
+
+    af::dim4 dims = in.dims();
+    af::dim4 strides = in.strides();
+    const Ti *inPtr = in.get();
+
+    Transform<Ti, To, af_add_t> transform;
+    Transform<uint, Tw, af_add_t> transform_weight;
+    kernel::MeanOp<To, Tw> Op(transform(inPtr[0]), transform_weight(1));
```

Comment:
You mean, no need to use `transform_weight`, right? `transform` is still needed to change Ti to To.